File: src/09.tree/10.543.go
package tree

func DiameterOfBinaryTree(root *TreeNode) int {
	var ans int = 0
	var dfs func(t *TreeNode) int
	dfs = func(t *TreeNode) int {
		if t == nil {
			return 0
		}

		l := dfs(t.Left)
		r := dfs(t.Right)

		if ans < l+r {
			ans = l + r
		}

		if l < r {
			return r + 1
		}
		return l + 1
	}
	dfs(root)
	return ans
}

File: src/09.tree/10.543.ts
import { TreeNode } from "./01.100";

export function diameterOfBinaryTree(root: TreeNode | null): number {
    let ans = 0;

    const dfs = (t: TreeNode | null): number => {
        if (!t) {
            return 0;
        }
        const l = dfs(t.left);
        const r = dfs(t.right);
        ans = Math.max(ans, l + r);
        return Math.max(l, r) + 1;
    };

    dfs(root);

    return ans;
}
